### PR TITLE
Make "Object Selection" order more coherent

### PR DESCRIPTION
**Pull request by karst**
This might take a short bit for people to get used to, but it's a lot more coherent and makes more sense in the end.
Better off doing this change before the release of 0.4.0 than after, so people will not be used to the old layout just yet.

I have done a poll on 03/12/2021 after doing a poll in the following discord servers:
Here's the poll results:
Original (A): https://cdn.discordapp.com/attachments/493167642853638164/916449391693418556/144671111-7cdf2e75-2252-4bc8-a59f-14b63ca8325d.png
Suggestion 1 (B): https://user-images.githubusercontent.com/1516893/144676419-b635ee89-4a3b-4055-b638-8b8385ef1463.png
Suggestion 2 (C) [In this PR]: https://user-images.githubusercontent.com/1516893/144680713-092c7f9f-e3a0-4745-9cdb-a8155c2b791a.png

- Marcel's server: 4A, 0B, 5C
- Deurklink's server: 1A, 2B, 13C
- Roller Coasters & Friends: 0A, 2B, 4C
- New Element: 0A, 10B, 1C
My vote: B

Totals (incl my own): 5A, 15B, 23C

This PR is for Order C. If you prefer another one. Or have suggestions on another order let me know.

### CODE REVIEW

**Comment by karst** [comment]
Oh boy, even though the order in the menus is fixed it now shows the wrong menus. Lemme figure this out

**Comment by karst** [comment]
Thanks MaxArceus for the help It's definitely a very learnful process! 😊 

I've tested this in game and it works! Should also make it easier to change the order around if any changes would be necessary again. Or probably if we'd ever decide to make UI customizable for the user.

File: src/openrct2-ui/windows/EditorObjectSelection.cpp [hunk 1]
```diff
@@ -125,27 +126,47 @@ struct ObjectPageDesc
...
+        {8, ObjectType::PathBits},
+        {9, ObjectType::Banners},
+		{10, ObjectType::Water},
+        {11, ObjectType::TerrainSurface},
+        {12, ObjectType::TerrainEdge},
+        {13, ObjectType::ParkEntrance},
+        {14, ObjectType::Station},
+        {15, ObjectType::Music},
+        {16, ObjectType::ScenarioText}
 };
```
Comment (spacek531) [inline src/openrct2-ui/windows/EditorObjectSelection.cpp:170]: This seems like it could be an array

**Comment by IntelOrca** [comment]
I am not sure, for me I would order it from landscape to ride.
* terrain, edge, water
* footpath, footpath surface, footpath railings, **footpath additions, footpath banners**
* scenery group, **scenery (small), scenery (large), scenery (walls)**
* rides, stations, music

Bold being hidden behind advanced.

File: src/openrct2-ui/windows/EditorObjectSelection.cpp (hunk 1 above)
Comment (karst) [reply src/openrct2-ui/windows/EditorObjectSelection.cpp:170]: How would I go about doing that? 

```
int TabOrderMap(ObjectType) {
       int x = 17;
       int16_t arr[x] = {ObjectType::Ride,
                         ObjectType::SceneryGroup,
                         ObjectType::SmallScenery,
                         ObjectType::LargeScenery,
                         ObjectType::Walls,
                         ObjectType::FootpathSurface,
                         ObjectType::FootpathRailings,
                         ObjectType::Paths,
                         ObjectType::PathBits,
                         ObjectType::Banners,
                         ObjectType::Water,
                         ObjectType::TerrainSurface,
                         ObjectType::TerrainEdge,
                         ObjectType::ParkEntrance,
                         ObjectType::Station,
                         ObjectType::Music,
                         ObjectType::ScenarioText};
}``` 

Like so?

**Comment by karst** [comment]
@IntelOrca like this?
image
image

**Comment by MaxMallon** [comment]
> @IntelOrca like this? image image

That order looks terrible to me. This game is ultimately about building rides, they should be first. Scenery and paths after that, then water and ground, and the remaining ones at the end, in my opinion.

**Comment by karst** [comment]
Perhaps we should swap those catagories around.
- rides, **stations**, **music**
- scenery group, **scenery (small)**, **scenery (large)**, **scenery (walls)**
- footpath surface, footpath railings, **footpath**, **footpath additions**, **footpath banners**, park entrance,
- terrain, edge, water,

File: src/openrct2-ui/windows/EditorObjectSelection.cpp (hunk 1 above)
Comment (karst) [reply src/openrct2-ui/windows/EditorObjectSelection.cpp]: Done :)

File: src/openrct2-ui/windows/EditorObjectSelection.cpp [hunk 2]
```diff
@@ -9,6 +9,7 @@
 
 #include <algorithm>
 #include <cctype>
+#include <map>
```
Comment (oli414) [inline src/openrct2-ui/windows/EditorObjectSelection.cpp:12]: I don't think you're using this anymore

File: src/openrct2-ui/windows/EditorObjectSelection.cpp [hunk 3]
```diff
@@ -125,29 +126,49 @@ struct ObjectPageDesc
...
+                         ObjectType::Paths,
+                         ObjectType::PathBits,
+                         ObjectType::Banners,
+                         ObjectType::Water,
+                         ObjectType::TerrainSurface,
+                         ObjectType::TerrainEdge,
+                         ObjectType::ParkEntrance,
+                         ObjectType::Station,
+                         ObjectType::Music,
+                         ObjectType::ScenarioText
```
Comment (oli414) [inline src/openrct2-ui/windows/EditorObjectSelection.cpp:169]: Is the scenario text object type also a page in the object selection?

File: src/openrct2-ui/windows/EditorObjectSelection.cpp (hunk 2 above)
Comment (karst) [reply src/openrct2-ui/windows/EditorObjectSelection.cpp:12]: Oh true, I will fix in next push

File: src/openrct2-ui/windows/EditorObjectSelection.cpp (hunk 3 above)
Comment (karst) [reply src/openrct2-ui/windows/EditorObjectSelection.cpp:169]: It's not as far as I know, but it's listed in every list so I thought of still including it just to be certain.

**Comment by karst** [comment]
image
image

What do you think of this?

**Comment by spacek531** [comment]
I'm not a fan of rearranging the scenery tabs in advanced mode

**Comment by oli414** [comment]
@spacek531 we already arranged them when some of the new tabs were introduced. If a lot of people don't like the current order then I think it's fine to change it, you'll get used to it after 2 uses

**Comment by oli414** [comment]
> when some of the new tabs were introduced. If a lot of people don't like the current order then I think it's fine to change it, you'll get used to it after 2 uses

I like grouping them by category where possible, but I do also feel like the frequency at which you typically use them is important to consider. I feel like the C suggestion in your initial post achieves that well.

File: src/openrct2-ui/windows/EditorObjectSelection.cpp [hunk 4]
```diff
@@ -125,29 +125,49 @@ struct ObjectPageDesc
...
+                         ObjectType::FootpathSurface,
+                         ObjectType::FootpathRailings,
+                         ObjectType::Paths,
+                         ObjectType::PathBits,
+                         ObjectType::Banners,
+                         ObjectType::Water,
+                         ObjectType::TerrainSurface,
+                         ObjectType::TerrainEdge,
+                         ObjectType::ParkEntrance,
+                         ObjectType::ScenarioText
```
Comment (Broxzier) [inline src/openrct2-ui/windows/EditorObjectSelection.cpp:168]: Always add a trailing comma when using arrays. That'll fix this weird indention.

**Comment by karst** [comment]
Anything else that I should change? Is it mayhaps an idea to get some community input on it, we can always change the order if people complain alot. But I do think this is a good layout :)

File: src/openrct2-ui/windows/EditorObjectSelection.cpp (hunk 4 above)
Comment (Broxzier) [reply src/openrct2-ui/windows/EditorObjectSelection.cpp]: Partly, yes. Now also apply clang-format here. It's currently wrapped between 'clang-format off/on' comments, so it gets ignored.

Comment (karst) [reply src/openrct2-ui/windows/EditorObjectSelection.cpp]: Aight cool, enabled it.

**Comment by oli414** [comment]
Do you have a screen of the order as the PR currently puts it?

**Comment by karst** [comment]
image

**Comment by Gymnasiast** [comment]
> Perhaps we should swap those catagories around.
> 
>     * rides, **stations**, **music**
> 
>     * scenery group, **scenery (small)**, **scenery (large)**, **scenery (walls)**
> 
>     * footpath surface, footpath railings, **footpath**, **footpath additions**, **footpath banners**, park entrance,
> 
>     * terrain, edge, water,

That seems like a good order to me.

**Comment by karst** [comment]
So that would only mean swapping around water with the park entrance. That's an easy fix. @oli414 do you want me to do that?

**Comment by oli414** [comment]
Yeah I'm cool with anything

**Comment by karst** [comment]
image

image

Alright, then this should be the final version. I have tested all tabs and they corrospond to the correct tab, so that's all good. Feel free to test again to be sure! Commit coming right up

**Review by oli414 — APPROVED** [review]

**Comment by oli414** [comment]
Might want to add a changelog entry btw

**Comment by karst** [comment]
That I have worked in develop as main branch before is gonna haunt me for life, init? 😂

**Comment by karst** [comment]
Well anyhow. I can't seem to get it to squash anymore and remove the old already previously merged commits from this branch. They are just annoying to be in this PR. You can ignore the Lint Commit Message failure as it's not even part of this PR. I am gonna have to completely clean up my develop branch after this is merged.

**Comment by karst** [comment]
Aight! Added the changelog entry! :)

File: src/openrct2-ui/windows/EditorObjectSelection.cpp (hunk 3 above)
Comment (Gymnasiast) [reply src/openrct2-ui/windows/EditorObjectSelection.cpp]: It’s not necessary any more, so I removed it for you.

**Review by Gymnasiast — APPROVED** [review]

**Pull request merged**